src/main/scala/io/dronekit/cloud/SignRequestForAWS.scala

```diff
@@ -0,0 +1,302 @@
+    val dateString = headers
+      .filter(_.lowercaseName() == "x-amz-date")
+      .map{header => header.lowercaseName() -> header.value().trim}
+      .map{case (name, values) => s"$values"}
+      .mkString("\n")
+    // how to choose something else if dateString is empty
+    if (dateString.length <= 1 ) {
+      getUTCTime()
+    } else {
+      dateString
+    }
+  }
+
+  //runs same algorithm as sign payload but on the canonicalRequest
+  def hashCanonicalRequest(canonicalRequest: String): String = {
+    val md = MessageDigest.getInstance("SHA-256")
+    md.update(canonicalRequest.getBytes("UTF-8"))
+    Hex.encodeHexString(md.digest())
+  }
+
+  //finds the relevant fields in the http request and returns the credential scope
+  def getCredentialScope(headers: Seq[HttpHeader], uri: Uri, region: String, service: String): String = {
+    val date = headers
+      .filter(_.lowercaseName() == "x-amz-date")
+      .map{header => header.lowercaseName() -> header.value().trim}
+      .map{case (name, values) => s"$values".split("T")(0)}
+      .mkString("")
+    date + "/" + region + "/" + service + "/" + "aws4_request"
+  }
+
+  //gets values for getSignatureKey and passes them in
+  //getSignatureKey(key: String, dateStamp: String, regionName: String, serviceName: String)
+  @throws(classOf[Exception])
+  def createSignatureKey(headers: Seq[HttpHeader], uri: Uri, key: String, region: String, service: String): Array[Byte] = {
+    val date = headers
+      .filter(_.lowercaseName() == "x-amz-date")
+      .map{header => header.lowercaseName() -> header.value().trim}
+      .map{case (name, values) => s"$values".split("T")(0)}
+      .mkString("")
+    getSignatureKey(key, date, region, service)
+  }
+
+  // got utc time for amz date from http://stackoverflow.com/questions/25991892/how-do-i-format-time-to-utc-time-zone
+  // got formatting from http://stackoverflow.com/questions/5377790/date-conversion
+  // formatting based on convention for amz signing
+  def getUTCTime(): String = {
+    val date = new Date()
+    val format1 = new SimpleDateFormat("yyyyMMdd")
+    format1.setTimeZone(new SimpleTimeZone(SimpleTimeZone.UTC_TIME, "UTC"))
+    val format2 = new SimpleDateFormat("HHmmss")
+    format2.setTimeZone(new SimpleTimeZone(SimpleTimeZone.UTC_TIME, "UTC"))
+    format1.format(date) + "T" + format2.format(date) + "Z"
+  }
+
+  // fixes all the uri encoding errors in adding the query String
+  def uriEncode(uriString: String): String = {
+    uriString.replace("+", "%20").replace(" ", "%20").replace("/", "%2F").replace(";", "%3B")
+  }
+
+
+//  def indexDocument(index: String, docType: String, docId: Option[String] = None, data: Map[String, String]): Future[HttpResponse] = {
+//    val documentKey = if (docId.isDefined) s"/${docId.get}" else ""
+//    val uri = s"/$index/$docType$documentKey"
+//    post(uri, data)
+//  }
+
+  def post(httpRequest: HttpRequest): Future[HttpResponse] = {
```

Comment:
This shouldn't be in here. If you need it for testing, then move it into the testing file.